src/ucp/core/ucp_worker.c

```diff
@@ -2345,32 +2364,63 @@ static unsigned ucp_worker_discard_uct_ep_progress(void *arg)
     return 1;
 }
 
-static int ucp_worker_discard_remove_filter(const ucs_callbackq_elem_t *elem,
-                                            void *arg)
+static int
+ucp_worker_discard_remove_filter(const ucs_callbackq_elem_t *elem, void *arg)
 {
-    if ((elem->cb != ucp_worker_discard_uct_ep_destroy_progress) &&
-        (elem->cb != ucp_worker_discard_uct_ep_progress)) {
+    if ((elem->arg != arg) ||
+        ((elem->cb != ucp_worker_discard_uct_ep_destroy_progress) &&
+         (elem->cb != ucp_worker_discard_uct_ep_progress))) {
         return 0;
     }
 
-    ucp_worker_put_flush_req((ucp_request_t*)elem->arg);
+    ucp_worker_discard_uct_ep_complete((ucp_request_t*)elem->arg);
     return 1;
 }
 
-static void ucp_worker_discarded_uct_eps_cleanup(ucp_worker_h worker)
+static void
+ucp_worker_discard_uct_ep_pending_count_purged(uct_pending_req_t *self,
+                                               void *arg)
+{
+    ++(*((unsigned*)arg));
+}
+
+static void ucp_worker_discard_uct_ep_purge_reqs(ucp_worker_h worker)
 {
+    ucs_hlist_head_t *reqs = &worker->discard_uct_ep_reqs;
+    ucp_request_t *req;
+    unsigned num_purged;
     uct_ep_h uct_ep;
 
-    ucs_callbackq_remove_if(&worker->uct->progress_q,
-                            ucp_worker_discard_remove_filter, NULL);
+    while (!ucs_hlist_is_empty(reqs)) {
```

Comment:
why need both hash and hlist?
also, do we really need the hash?

Reply:
we need:
- hlist for convenient robust iterations khash doesn't provide it, we can implement it though, e.g.
```
while (kh_size(<hash>) != 0) {
    kh_foreach(... {
        <delete key from hash>
        break;
    });
}
```
- hash for search

> also, do we really need the hash?
https://github.com/openucx/ucx/blob/6d3fcfcf446b25cc15c63438a20c2604f2a7e015/src/ucp/core/ucp_worker.c#L625

Reply:
1. better use kh_foreach than introducing extra code/fields/data structure. this is slowpath.
2. right - thanks for reminder

Reply:
> * better use kh_foreach than introducing extra code/fields/data structure. this is slowpath.

```
while (kh_size(<hash>) != 0) {
    kh_foreach(... {
        <delete key from hash>
        break;
    })
}
```
is this way ok?

Reply:
can we just avoid "delete key from hash" during the loop?

Reply:
done

Reply:
> can we just avoid "delete key from hash" during the loop?

we can provide some flag though UCP requests, but I don't think that this a good idea.

Reply:
how about moving to temp local list?

Reply:
> how about moving to temp local list?

makes sense - done